doc/design.md

```diff
@@ -0,0 +1,175 @@
+
+### Placeholder types
+
+`task_handle`: A identifier for a running task, unique across the set of currently registered tasks. The exact type this will map to is not yet defined, but likely to be a either a `string` or an `int64`. The data it encodes will *not* be a simple unix PID, as the API server will have to track the state of processes that have exited, and the system may recycle the PIDs in the lifetime of the API server.
+
+`error`: Some data indicating an error condition, e.g. "No such binary", "Access Denied", "No such task", etc
+
+`range`: Some type indicating the range of bytes in a stream that the caller is interested in.
+
+`task_status`: An enumeration of values indicating the current state of a task, e.g. `running`, `exited`.
+
+### Starting a task
+
+`launchTask(binary_path string, working_dir *string, environment: map[string]string, args ...string) (task_handle, error)`
+
+Starts an arbitrary process and returns a handle that can be used to interact that process. The launched process will run under, and inherit the permissions of, the account of the user running the API server. No shell variable substitution will be performed.
+
+**Arguments**
+-  `binary_path`: fully-qualified path to the binary to launch
+-  `working_dir`: an optional working directory for the process to start in
+-  `environment`: a map of key/value pairs vars to add to the environment of the process
+-  `args`: a list of command-line arguments to be passed to the process on startup
+
+**Returns**
+Returns either a `task_handle` that identifies the launched task, or an `error` indicating a failure condition. Note that the system *may* recycle the `task_handle` of a task that has been freed.
+
+### Querying a task
+
+`queryTask(task: task_handle, stdout_range: *range, stderr_range *range) (task_status, []byte, []byte, error)`
+
+Queries the state of a given task, returning the `task_status`, and optionally returning the task's `stdout` and `stderr` streams.
+
+Assume that the `stdout` and `stderr`  `range`s have a sentinel "end" value that indicates the caller wants all the remaining bytes in the stream.
+
+**Arguments**
+-  `task`: The `task_handle` that uniquely identifies a running task. Note that querying a non-existent task, or a task that was not started by the caller, is considered an error.
+-  `stdout_range`: An optional `range` indicating the range of bytes from the task's stdout stream that the caller wants. A `nil` value indicates that the caller is not interested in receiving the logs with this call.
+-  `stderr_range`: An optional `range` indicating the range of bytes from the task's stderr stream that the caller wants. A `nil` value indicates that the caller is not interested in receiving the logs with this call.
```

Comment:
I would simply, just get all the logs at once. However, you may want a different RPC endpoint for that.

Reply:
Fair enough. Will refactor that in an upcoming commit.

Reply:
Done